magenta-lib/src/main/scala/magenta/tasks/gcp/GCSTasks.scala

```diff
@@ -102,6 +115,75 @@ case class GCSUpload(
   }
 
   private def cacheControlLookup(fileName:String) = cacheControlPatterns.find(_.regex.findFirstMatchIn(fileName).isDefined).map(_.value)
+
+  private def getCurrentObjectsForDeletion(storage: Storage) : List[StorageObject] = {
+
+    val objectsInThisDeploy = transfers.map(_.target).toList
+
+    def getAllObjectsForDirectory(query: Storage#Objects#List,  foundSoFar: List[StorageObject] = List.empty): List[StorageObject] = {
+
+      val listResults = query.execute
+      val currentPageItems = Option(listResults.getItems).map(resultsSet => resultsSet.asScala).getOrElse(List.empty)
+      val allItemsFoundSoFar = foundSoFar ++ currentPageItems
+
+      Option(listResults.getNextPageToken) match {
+        case Some(token) => getAllObjectsForDirectory(query.setPageToken(token), allItemsFoundSoFar)
+        case None => allItemsFoundSoFar
+      }
+    }
+
+    def allObjectsInMatchingDirectories(directoriesToPurge: List[String], itemsSoFar: List[StorageObject] = List.empty): List[StorageObject] =
+        directoriesToPurge match {
+          case Nil => itemsSoFar
+          case head :: tail =>
+            val gcsQuery = storage.objects()
+             .list(gcsTargetBucket.name)
+             .setPrefix(head)
+            val allItemsForThisDirectory = getAllObjectsForDirectory(gcsQuery, itemsSoFar)
+            allObjectsInMatchingDirectories(tail, allItemsForThisDirectory )
+        }
```

Comment:
How does this work if it is a directory (rather than a file) in the storage object? Do we need to cover that case..? maybe not

Reply:
Yeah the v1 google api sucks a bit. This does what we want. If the directory is `dags/dags`(Like [Here](https://console.cloud.google.com/storage/browser/europe-west2-datatech-airfl-fdf1e4b5-bucket/dags/dags?project=datatech-platform-code&pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&prefix=&forceOnObjectsSortingFiltering=false)), this gives all objects, recursively under the folder in head. I don't think folders are considered as storage options. You can sae this by running this code by itself 

```
object GcpBucketReaderTwo extends App{

  lazy val httpTransport: ApacheHttpTransport = GoogleApacheHttpTransport.newTrustedTransport
  //lazy val jsonFactory = JacksonFactory.getDefaultInstance()
  lazy val jsonFactory = JacksonFactory.getDefaultInstance()
  private val gcsTargetBucket = "europe-west2-datatech-airfl-fdf1e4b5-bucket"


  authorize.map{
    credentials =>
      println("Got credentials")
      val storage = new Storage.Builder(httpTransport, jsonFactory, credentials ).build()

     val results = allObjectsInMatchingDirectories(storage, List("dags/dags", "dags/gssuardian") )
     results.foreach(s => println(s"**${s.getName}**"))
  }

  def allObjectsInMatchingDirectories(storage: Storage, directoriesToPurge: List[String], itemsSoFar: List[StorageObject] = List.empty): List[StorageObject] =
    directoriesToPurge match {
      case Nil => itemsSoFar
      case head :: tail =>
        val gcsQuery = storage.objects()
          .list(gcsTargetBucket)
          .setPrefix(head)
        val allItemsForThisDirectory = getAllObjectsForDirectory(gcsQuery, itemsSoFar)
        allObjectsInMatchingDirectories(storage, tail, allItemsForThisDirectory )
    }



  @tailrec
  def getAllObjectsForDirectory(query: Storage#Objects#List, foundSoFar: List[StorageObject] = List.empty): List[StorageObject] = {

    println(s"Excecusted: ${foundSoFar.size}")
    val results = query.execute
    println("Excecusting")
    val currentPageItems = Option(results.getItems).map(rs => rs.asScala).getOrElse(List.empty)
    val allItemsSoFar = foundSoFar ++ currentPageItems


    val token = results.getNextPageToken
    Option(token) match {
      case None => allItemsSoFar
      case Some(nextPageToken) => getAllObjectsForDirectory(query.setPageToken(nextPageToken), allItemsSoFar)
    }
  }
```

Reply:
OK there was an issue here which I've fixed in the last commit. I don't think I can delete folders tho